extensions/ql-vscode/src/vscode-tests/no-workspace/remote-query-history.test.ts

```diff
@@ -0,0 +1,252 @@
+    expect(openTextDocumentSpy).to.have.been.calledOnce;
+
+    const uri: Uri = openTextDocumentSpy.getCall(0).args[0];
+    expect(uri.scheme).to.eq('codeql');
+    const params = new URLSearchParams(uri.query);
+    expect(params.get('isQuickEval')).to.eq('false');
+    expect(params.get('queryText')).to.eq(rawQueryHistory[0].remoteQuery.queryText);
+  });
+
+  describe('AnalysisResultsManager', () => {
+    let mockCredentials: any;
+    let mockOctokit: any;
+    let mockLogger: any;
+    let arm: AnalysesResultsManager;
+
+    beforeEach(() => {
+      mockOctokit = {
+        request: sandbox.stub()
+      };
+      mockCredentials = {
+        getOctokit: () => mockOctokit
+      };
+      mockLogger = {
+        log: sandbox.spy()
+      };
+      sandbox.stub(Credentials, 'initialize').resolves(mockCredentials);
+
+      arm = new AnalysesResultsManager(
+        {} as ExtensionContext,
+        path.join(STORAGE_DIR, 'queries'),
+        mockLogger
+      );
+    });
+
+    it('should avoid downloading an analysis result', async () => {
+      const publisher = sandbox.spy();
+      const analysisSummary = remoteQueryResult0.analysisSummaries[0];
+      await arm.downloadAnalysisResults(analysisSummary, publisher);
+
+      // Should not have made the request since the analysis result is already on disk
+      expect(mockOctokit.request).to.not.have.been.called;
+
+      // result should have been published twice
+      // first time, it is in progress
+      expect(publisher.getCall(0).args[0][0]).to.include({
+        nwo: 'github/vscode-codeql',
+        status: 'InProgress',
+        // results: ... avoid checking the results object since it is complex
+      });
+
+      // second time, it has the path to the sarif file.
+      expect(publisher.getCall(1).args[0][0]).to.include({
+        nwo: 'github/vscode-codeql',
+        status: 'Completed',
+        // results: ... avoid checking the results object since it is complex
+      });
+
+      // result should be stored in the manager
+      expect(arm.getAnalysesResults(rawQueryHistory[0].queryId)[0]).to.include({
+        nwo: 'github/vscode-codeql',
+        status: 'Completed',
+        // results: ... avoid checking the results object since it is complex
+      });
+    });
+  });
+
+  // Since this test changes the state of the query history manager, we need to copy the original
+  // to a temporary folder where we can manipulate it.
+  function copyHistoryState() {
+    fs.ensureDirSync(STORAGE_DIR);
+    fs.copySync(path.join(__dirname, 'data/remote-queries/'), path.join(tmpDir.name, 'remote-queries'));
+
+    // also, replace the files with "PLACEHOLDER" so that they have the correct directory
+    for (const p of walk(STORAGE_DIR)) {
+      replacePlaceholder(path.join(p));
+    }
+  }
+
+  function deleteHistoryState() {
+    fs.removeSync(STORAGE_DIR);
+  }
+
+  function* walk(dir: string): IterableIterator<string> {
```

Comment:
Nice! At first I was thinking it might be an overkill to have an a generator for tests, but there's no harm in having it and puts us in a good position if we want to test a lot/big files.

Reply:
I should probably move this to a shared location since it might be useful in other places.